### ISSUE TITLE
Explorer Plugin Enhancement: Display Target Path for .lnk Files (in addition to Location Path)

### ISSUE THREAD

**Issue by EnigmaSolved** [Flow-Launcher/Flow.Launcher#2560]
I think it could be useful in general (as well as for my specific use case, described below) if `.lnk` shortcut files would display the Target Path (probably with the prefix `Target:`) below the currently displayed Location Path.

This is particularly useful for the following scenario: I have my Recents folder (`%APPDATA%\Microsoft\Windows\Recent Items`) indexed by Everything, and I have this folder added to my Quick Access Links in Flow. This allows me to type `recent` into Flow, then `Tab` to expand it and `>` to search a list of recent files and folders (and, of course, a Custom Query Hotkey makes all this even quicker!). But since the Windows Recent Items folder is a special folder it is full only of `.lnk` files and thus the path information is useless in this case. 

Now, while it would be nifty to add special handling for this unique folder (such that instead of displaying File/Folder Location Path at all we only displayed the `.lnk` Target), the easier and more universally useful thing would be an option to _additionally_ display the Target Path beneath Location Path for all `.lnk` files. This would allow me to distinguish between multiple Target files of the same name (but different Target locations) and more quickly isolate what I am looking for. :)

Thanks for considering my request!

Edit: Meant to add that I wonder if this request overlaps at all with https://github.com/Flow-Launcher/Flow.Launcher/issues/2390, at least in terms of some of the backend code?

**Comment by jjw24** [comment]
>Edit: Meant to add that I wonder if this request overlaps at all with #2390, at least in terms of some of the backend code?

That's fine, it's more of a dependency, need to get that code done to get the target path before able to display. Displaying would be easy change.

**Comment by EnigmaSolved** [comment]
> That's fine, it's more of a dependency, need to get that code done to get the target path before able to display. Displaying would be easy change.

Excellent. That's great to hear!

**Comment by EnigmaSolved** [comment]
> > Edit: Meant to add that I wonder if this request overlaps at all with #2390, at least in terms of some of the backend code?
> 
> That's fine, it's more of a dependency, need to get that code done to get the target path before able to display. Displaying would be easy change.

@jjw24 Just bumping this now that #2390 / #2508 is done in hopes that this can get added into the 1.18.0 release (though I will totally understand if it needs to be pushed to a later release). :) Thank you for all of your hard work!

**Comment by EnigmaSolved** [comment]
Would still love to see this implemented! :)

**Comment by EnigmaSolved** [comment]
Thanks much for reopening -- though it is a little thing, I think it will be a really useful enhancement! :)

**Comment by EnigmaSolved** [comment]
Just giving this a bump to keep it alive! :)

**Comment by EnigmaSolved** [comment]
Just giving this a bump. :)

**Comment by EnigmaSolved** [comment]
Would still love to see this implemented (whenever you all can get to it)! 😁

**Comment by onesounds** [comment]
Hello. I just saw this post now.
Upon checking, I see that this folder is where the Start Menu stores the data for its recent items list.

Currently, Flow cannot display multiple lines in the subtitle area, so adding the target path of .lnk files as an additional subtitle line — as you suggested — is not feasible.
An alternative option, such as “show target path instead of location path for all .lnk files”, is technically possible. However, this behavior would only be useful for this specific use case, while in most cases it would introduce unnecessary noise. In general, from a UI perspective, we prefer not to display the target path for .lnk files as it's usually not relevant.

Therefore, I believe there are two possible solutions:

1. Create a dedicated plugin (Recently Used)

- This plugin would allow users to add specific folders, and handle .lnk files within those folders differently, showing their target paths directly.

- With proper action keywords or configuration, this could essentially work the same way as your current workflow.

-  This feature would only be provided to users who specifically need it.

2. Add special handling for this folder

- First, I need to verify whether this folder path is consistent across all Windows versions.

- Then, for .lnk files within this folder, we can simply display the target path instead of the location path

From a development perspective, I think option 2 would be more convenient, but we need to verify whether there’s any performance impact when implementing it.
Also, if we allow users to easily add this folder to Flow via a simple checkbox, it could provide additional convenience for other users as well. I may consider this option for inclusion in the core.

Since you’ve been waiting for a while, I’ll prioritize this issue as my top task for this week.
Once the feature is implemented, I’ll share a test build for you to try.

**Comment by EnigmaSolved** [comment]
I'm so grateful for your willingness to look into this! And I agree with all of your points. :) I think Option 2 and potential inclusion in core could make a lot of sense -- I have used a dedicated program (that is terribly out of date) for this purpose in the past, so having it built in to Flow would be awesome.

Let me know anything you need from me, thoughts, feedback, testing, etc -- happy to assist in whatever ways I can!

FYI in case it effects things: I use a portable installation [synced to OneDrive](https://github.com/Flow-Launcher/Flow.Launcher/issues/2684#issuecomment-2525189725) so I can use it across multiple computers (all of which are Windows 10/11 at this point, and soon to all be Win11). 

Edit: Also, it has _seemed_ to me that the Recent Items location has been consistent in Windows for a while, but I can't say for absolute certain. If you can't find a definitive answer perhaps you could provide a way for folks to override that folder path location if need be (maybe just via the settings file initially so you don't have to bother with additional UI). My gut feeling (for what it is worth) is that the location is probably fairly stable.

**Comment by onesounds** [comment]
[Flow.Launcher.Plugin.RecentlyUsed.zip](https://github.com/user-attachments/files/20757644/Flow.Launcher.Plugin.RecentlyUsed.zip)

Hi,
I’ve packaged the plugin — please extract it into your plugin directory and test it. The default action keyword is recent, and for now, only the basic functionality has been implemented. When you press Enter on a folder, it simply opens. If you use the autocomplete key (usually Tab), the query is transformed into the full path so that it can be handled by the Explorer plugin. Please let me know if this function meets the requirements.

From my tests, I found that this feature feels more like a database that records history entries as .lnk files, rather than something intended for use like a regular file explorer. (In fact, after mimicking your usage, I noticed that some of the paths couldn't even be specified via normal Windows file dialogs — I suspect you manually edited configuration files to assign those paths.) Internally, there are also special folders for handling URIs (like links to open OneNote files).

Therefore, I believe using these paths with the Explorer plugin is more of a workaround, and directly implementing your request as-is would likely result in a feature that only applies to your specific use case. (We actually had quite a bit of discussion on whether to simply display the .lnk target paths or to build a separate plugin for it.)

Since I think this functionality, if properly designed, might be useful for other users as well, I decided to implement it as a separate plugin.
Please test it and let me know if there are any additional features you’d like to see. (To be honest, I’m not even sure if it will work as expected — this is my first time building a plugin.)

At the moment, I’m considering adding features like:

- File deletion

- Clear entire folder (removing all previously used app history)

- Showing a message when the system setting is disabled

- Copy file name or path via context menu

- Displaying information like “2 hours ago” as tooltip or preview

If you have any further ideas, please feel free to share.

**Comment by EnigmaSolved** [comment]
You are amazing!! 👏 I've only just done a little initial testing, but so far it overall seems great. I have a few initial thoughts and observations (including one little possible bug). I'm about to be offline for a few hours, but should be able to play with this more later today, and will just in general start making more use of it so I can see what I bump into, what other ideas/feedback occur to me, etc.

> (In fact, after mimicking your usage, I noticed that some of the paths couldn't even be specified via normal Windows file dialogs — I suspect you manually edited configuration files to assign those paths.) 

I would love to hear more about this, just so I can confirm (or not) whether I've done anything special on my end.

I do see, though, some weird examples (which I've not tried to open) that aren't exactly folder or file paths.

All of the additional features you mentioned sound good, but I personally would value the following two the most:
> Copy file name or path via context menu

> Displaying information like “2 hours ago” as tooltip or preview

I have a network share (eg, \\NAS_ROOT\SHARE1) in my Recents list. I appears and functions properly in Windows Explorer (eg, SHARE1 and clicking on it goes to the share), but in the plugin I only see the network icon and everything else is blank, and clicking on the item does not open the location (which makes sense, since it seems like Flow lacks that info in this moment).

Another thing is that if I type `recent ` it takes a bit (5-10 seconds) for the list to populate. Personally for my use, I would prefer that it default to showing the list as soon as I type `recent` as I'm typically going to be trying to remember something and thus wanting to browse the list visually anyway.

**Comment by onesounds** [comment]
- When I try to add the path to Quick Access in Explorer, the path doesn’t appear in the folder selection dialog, so I have to manually type it in. (Some specific paths behave like this.) If you haven’t made any changes on your side, it’s probably an issue on my end.

- I'm curious whether you use this feature from the Start Menu. In my tests, I accessed the path and saw lnk entries created, but some of them didn’t show up in the Start Menu. (It seems that the Start Menu applies some filtering.) In such cases, the entries appear in Flow but not in the Start Menu. For example, I noticed this behavior for some system settings pages I accessed.

- I’ll check the network paths and result list loading (I think caching should help with any performance issues). Roughly how many files exist in that location?

- If there are any unexpected files being shown, or file types that don’t open properly, please let me know (even roughly) — that would be helpful.

I’ll stop here for today as well. I’m keeping an eye on this thread, so feel free to leave feedback anytime.

**Comment by EnigmaSolved** [comment]
I have only used it from the Quick Access list. I think I added it by first finding it manually and then right-clicking to "Pin to Quick Access" which then just opens the folder in Explorer.

I'll start making a list of weird files that don't open.

**Comment by onesounds** [comment]
This path can contain various types of files, but it's a bit different from what actually appears in the "Recent items" section of the Start Menu or File Explorer. Items that don't appear there—such as opening a non-file path like C:\download\—can still be added to this folder, but they won’t show up in the Start Menu's recent items.
Would it be okay to hide those items that don't appear in the Start Menu's list?
In other words, should I implement this feature based on what shows up in the Start Menu's "Recent items"?

**Comment by EnigmaSolved** [comment]
My initial thought is that I'd be fine with that -- I mainly use it for finding recent normal file/folder items, which should show up fine with filtering like you're describing. So I'd say go ahead and move forward with that plan.

Also, this might be no longer needed, but here's the few odd items I noticed so far that don't open normally in that folder:
```
.exportedUI (items with this file extension)

::{A0D4CD32-5D5D-4F72-BAAA-767A7AD6BAC5}  This one is a Control Panel item for Outlook Mail settings (I can provide a screenshot if needed)

ms-gamingoverlay:///
```

**Comment by EnigmaSolved** [comment]
>I think caching should help with any performance issues

It randomly just occurred to me that with regards to adding some caching that I should reiterate that I use a portable installation [synced to Microsoft OneDrive](https://github.com/Flow-Launcher/Flow.Launcher/issues/2684#issuecomment-2525189725) so I can use it across multiple computers (all of which are Windows 11 at this point). At the very least, let me know where the cache file is located so that I can add it to my script. 

I have a little script that runs when I unlock my computer and which deletes `UWP.cache` and `Win32.cache` before reloading Flow so that those get re-built for the current machine. It is clunky, but it works. So I'll probably need to do something similar with any cache for Recent files/folders. And ideally the plugin would similarly rebuild the Recent cache on launching Flow.

**Comment by onesounds** [comment]
[Flow.Launcher.Plugin.RecentlyUsed.zip](https://github.com/user-attachments/files/20923850/Flow.Launcher.Plugin.RecentlyUsed.zip)

I've made a slightly improved version. I fixed a bug related to files that couldn't be executed, and excluded some unnecessary file types from the result list. As for the cache implementation, I'm still figuring out the best approach. I'm already having some difficulty with the basic caching mechanism itself, so it might take a bit of time.

And I think it's more helpful if you change your requests as follows:

**Your Request**: "Please add a setting to show the target path of an .lnk file."
**Correct Request**: "Please create a plugin that displays the system's recently used items list."

As I mentioned, the 'recent items' path isn't meant to be used with a standard file manager; it should be seen as a kind of database. (You can probably tell this path isn't designed for regular users.) Trying to use it as an Explorer plugin is very tricky and uncommon. (You likely know that very few users open this path with a file manager.) Requests like this make it hard for developers to understand the real need and decide whether to proceed. Displaying the .lnk path provides no benefit outside of this specific path, and since most users won't use it, it ends up being a specialized feature just for you. Instead of this suggestion, it would be much better to explain the necessity and usefulness of the recent list and the programs you use it with.

**Your Request**: "Please tell me the path so I can clear the cache with my script."
**Correct Request**: "Please improve the system so that each PC can use its own cache in a synchronized portable environment."

This is how it should be. You're an advanced user who can write scripts, and your request is just a temporary fix because the current functionality is lacking. Objectively, the core issue is that the cache usage structure needs to be improved in synchronized environments. From my perspective, simply telling you the path to clear the cache with a script or having the system rebuild it every time for a portable environment isn't an appropriate solution. Like the .lnk path suggestion, it's a solution that would be meaningless to other users.

In my experience, almost everyone, not just you, fails to explain the necessity of a feature and a suitable alternative. They just ask for what they think is the solution. Let's use a car as an example. If the steering wheel doesn't turn left, a user might submit a request on GitHub saying, "Please add a button to turn left." Then the developer adds "a button for the car to turn left." In my opinion, the user should report, "The steering wheel does not turn left," and the developer should fix the steering wheel. From this perspective, it's always best to explain the need for the feature and the alternative you are currently using.

I also need to check a few things:

- Other system locations that use this path (File Explorer, Start Menu) only show "files," not "folders." (Currently, displaying them as an option.) Do you believe "folders" should also be displayed?
- How many .lnk files are in that folder, and how old is the oldest file you've opened before?

**Comment by EnigmaSolved** [comment]
>Other system locations that use this path (File Explorer, Start Menu) only show "files," not "folders." (Currently, displaying them as an option.) Do you believe "folders" should also be displayed?

Ah good question! Folders are actually pretty important to me (along with recent files). I suppose eventually you could have two checkbox options for whether folks want Files displayed and Folders displayed (and default to both being checked). Personally, I think it makes sense to just include both, but perhaps in the future some folks will only care about one or the other.

>How many .lnk files are in that folder, and how old is the oldest file you've opened before?
On one of my computers (as of today) I'm seeing 150 .lnk files, with the oldest being April 1, 2025.

I'll let you know if I run into anything with the updated version!

I agree with your points (about providing more thorough explanations and requests) 100%. I used to do that a lot more (and sometimes still do), but I have found it to be rarely effective, unfortunately (which is not at all a statement about the Flow community in particular, but rather a much broader statement of my general experiences). People seem to get overwhelmed by the details and/or it takes years (if ever) for something to actually get implemented. So pragmatically I have defaulted much of the time to whatever is the smallest bite/request that I think might facilitate what I'm actually wanting (so that I have a chance of implementing something relatively quickly). But you're totally correct on all of your points. I've just allowed discouragement to train me to going the route of less effort much of the time (though I'm not justifying that as a better, but just my concession to the perceived limits of others I have often run into). I'm going to update the title for this thread shortly, though. I also confess I had not had it even as a possibility in my mind to hope for the level of functionality that you have already provided via plugin, so it didn't even occur to me to ask for that much originally (and I'm even more so immensely grateful for your response to the underlying issue). 😊

>Correct Request: "Please improve the system so that each PC can use its own cache in a synchronized portable environment."

That is exactly what I ultimately would like to see. And I think the linked thread about better portable support talks about that some (if I remember correctly). I suppose another thing that happens in me psychologically is that I don't want to be demanding, especially in an environment like this where you all are giving your time so generously. But you are correctly discerning that improving the portability functionality, especially regarding various cache handling is something I would very much like to see in the future.

I am grateful to be encountering you and your desire for more detail, more thorough explanations and requests!

**Comment by EnigmaSolved** [comment]
Newest version seems broken on my end (ie, it gives the error `Recently Used: Failed to Respond!`).

Log: [2025-06-26.txt](https://github.com/user-attachments/files/20928158/2025-06-26.txt)

```
Flow Launcher version: 1.19.4
OS Version: 26100.4061
IntPtr Length: 8
x64: True

Python Path: C:\Users\Sean\OneDrive\Apps\FlowLauncher\app-1.19.4\UserData\Environments\Python\PythonEmbeddable-v3.11.4\pythonw.exe
Node Path: 
Date: 06/26/2025 10:24:30
Exception:
Recently Used Exception: 
Websites: http://flowlauncher.com
Author: onesounds
Version: 1.0.0
Flow.Launcher.Core.ExternalPlugins.FlowPluginException: Method not found: 'Void Flow.Launcher.Plugin.Result.set_AddSelectedCount(Boolean)'.
 ---> System.MissingMethodException: Method not found: 'Void Flow.Launcher.Plugin.Result.set_AddSelectedCount(Boolean)'.
   at Main.Query(Query query)
   at Flow.Launcher.Plugin.IPlugin.<>c__DisplayClass3_0.<Flow.Launcher.Plugin.IAsyncPlugin.QueryAsync>b__0() in /_/Flow.Launcher.Plugin/Interfaces/IPlugin.cs:line 35
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Flow.Launcher.Core.Plugin.PluginManager.<>c__DisplayClass28_0.<<QueryForPluginAsync>b__0>d.MoveNext() in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 248
--- End of stack trace from previous location ---
   at Flow.Launcher.Infrastructure.Stopwatch.DebugAsync(String message, Func`1 action) in C:\projects\flow-launcher\Flow.Launcher.Infrastructure\Stopwatch.cs:line 34
   at Flow.Launcher.Core.Plugin.PluginManager.QueryForPluginAsync(PluginPair pair, Query query, CancellationToken token) in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 247
   --- End of inner exception stack trace ---
   at Flow.Launcher.Core.Plugin.PluginManager.<>c__DisplayClass28_1.<QueryForPluginAsync>b__1(ActionContext _) in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 276
   at Flow.Launcher.Plugin.Result.ExecuteAsync(ActionContext context) in /_/Flow.Launcher.Plugin/Result.cs:line 246
   at Flow.Launcher.ViewModel.MainViewModel.OpenResultAsync(String index) in C:\projects\flow-launcher\Flow.Launcher\ViewModel\MainViewModel.cs:line 389
   at CommunityToolkit.Mvvm.Input.AsyncRelayCommand.AwaitAndThrowIfFailed(Task executionTask)
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
```

**Comment by onesounds** [comment]
I have checked the problem. I'll share it as soon as the revised version comes out. Please wait for about a day. 😅

**Comment by EnigmaSolved** [comment]
> I have checked the problem. I'll share it as soon as the revised version comes out. Please wait for about a day. 😅

Take as long as you need! 😁

**Comment by onesounds** [comment]
[Flow.Launcher.Plugin.RecentlyUsed.zip](https://github.com/user-attachments/files/20928808/Flow.Launcher.Plugin.RecentlyUsed.zip)
Hotfix. Try this.

**Comment by EnigmaSolved** [comment]
> [Flow.Launcher.Plugin.RecentlyUsed.zip](https://github.com/user-attachments/files/20928808/Flow.Launcher.Plugin.RecentlyUsed.zip) Hotfix. Try this.

That works! And loads very fast now!

**Comment by EnigmaSolved** [comment]
So far it is loading results quite promptly, so caching may be unnecessary. :)

One thing though, the sorting seems a little off: I think we want to be sorting based on "Date Modified" as I think that should yield most recently items first (ie, top of the list). Although, it just occurred to me that sorting could possibly be impacted by the Flow functionality of moving more frequently clicked items to the top?? Though I can't recall if I have been selecting any particular item repeatedly or not, but the item I am seeing at the top (which should be further down) is an item I have added to my Explorer plugin Quick Access List.

**Comment by onesounds** [comment]
In my environment, caching does seem necessary. I experienced some slow loading issues. I’ll test and address this on my end.

As for the sorting, I think your guess is probably correct. I did apply sorting by date, but I vaguely remember there being some logic in the result rendering that could be affected by score. I’m not entirely sure, so I’ll check if it’s possible to bypass that.

**Comment by EnigmaSolved** [comment]
> In my environment, caching does seem necessary. I experienced some slow loading issues. I’ll test and address this on my end.

I did notice a lengthy delay once recently on one of my computers (though it was then fine for subsequent calls) -- I use Recent intermittently, so I'm slow at collecting observational data, but will let you know as I observe more.

**Comment by onesounds** [comment]
[Flow.Launcher.Plugin.RecentlyUsed.zip](https://github.com/user-attachments/files/21116986/Flow.Launcher.Plugin.RecentlyUsed.zip)

Here's an updated version with a few new features

- Caching Implemented: The application now utilizes caching, with different cache files saved and loaded based on the PC name.
- New Copy Options: You can now Copy the filename. / Copy the full path. /Open the file's location directly.

**Comment by EnigmaSolved** [comment]
That is a brilliant idea regarding the computer-specific caching! I'm curious, what triggers and/or how often does a cache refresh happen?

I'll let you know if I run into any issues! It is looking great, though!

**Comment by onesounds** [comment]
Plugin Cache Update Logic
- During Plugin Initialization

  - When Flow Launcher starts or the plugin is loaded, Task.Run(UpdateCache) is called at the end of the Init method to generate and refresh the cache in the background.

- When the 'Recent Items' Folder Changes

  - A FileSystemWatcher monitors Windows' "Recent Items" folder in real time. When a shortcut file (.lnk) is created, modified, deleted, or renamed, the OnRecentFolderChanged event is triggered. This event starts _updateTimer, which waits 500 milliseconds before calling UpdateCache. The delay minimizes system load by coalescing rapid, successive file changes into a single cache update.

- When Plugin Settings (Show Folder) Are Changed

**Comment by EnigmaSolved** [comment]
That cache update logic is great!

>Caching Implemented: The application now utilizes caching, with different cache files saved and loaded based on the PC name.

I don't know how feasible it might be, but it would be awesome to see this same functionality ported into core for any cache files (so that I'm not doing my hacky workaround of deleting cache files all the time for portable mode used across multiple computers).

**Comment by EnigmaSolved** [comment]
I continue to appreciate this plugin, @onesounds! I have an idea for an enhancement: It would be nifty to be able to display the `recent` view on the Flow Launcher "Home Page" view, without needing the Action Keyword. I don't use any of the other Home Page views, so for me effectively this would mean opening Flow would provide me with instant access to my recently used files.

Now admittedly this idea requires some further ironing out regarding things like:
1. Should the Action Keyword (eg, `recent`) be prepopulated, and if not, what do we want to happen when the user starts typing (ie, filter the Recents list, or perform a normal search)? My inclination is to _not_ have the Action Keyword present, as then I'll have to backspace it all the times where I want to just do a search. And likewise, it seems to me that typing should clear the Recents list, just like any Home Page view is cleared as soon as the user starts typing.
2. In light of the above, should there be a way to still filter the Recents list from this Home Page view, and if so, how? I don't have an immediate idea for how to easily do this. Worst case the user just has to type their Action Keyword first like normal and then can start filtering the Recents.

Let me know if this is an enhancement you're interested in adding, and if so we can further iron out the functionality. :)
